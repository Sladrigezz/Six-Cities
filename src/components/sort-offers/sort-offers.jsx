import React from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';

import {sortOptions, SortTypes} from '../../constants/constants.js';
import {getActiveSortType, getIsSortOpen} from '../../selectors/sort.js';
import {ActionCreator} from '../../reducers/index.js';

export const SortOffers = (props, {activeSortType, isOpen, setSortType, sortFilteredOffers}) => {
  const handleClick = () => {
    props.toggleSort();
    document.removeEventListener(`click`, handleClick);
  };

  const onSortClickHandler = () => {
    props.toggleSort();
    document.addEventListener(`click`, handleClick);
  };

  const onSortChangeHandler = (e) => {
    setSortType(e.currentTarget.id);
    sortFilteredOffers();
  };

  return (
    <form className="places__sorting" action="#" method="get">
      <span className="places__sorting-caption">Sort by</span>
      <span onClick={onSortClickHandler} className="places__sorting-type" tabIndex="0">
        Popular
          <svg className="places__sorting-arrow" width="7" height="4">
          <use xlinkHref="#icon-arrow-select"></use>
        </svg>
      </span>
      <ul className={`places__options places__options--custom ${isOpen ? `places__options--opened` : ``}`}>
        {sortOptions.map((item, i) => {
          const SortTypesValues = Object.values(SortTypes);
          const currentSortType = SortTypesValues[i];
          const isActive = activeSortType === currentSortType;
          return (
            <li
              className={`places__option ${isActive ? `places__option--active` : ``}`}
              tabIndex={i}
              key={item}
              id={currentSortType}
              onClick={onSortChangeHandler}
            >
              {item}
            </li>
          );
        })}
      </ul>
    </form>
  );
}


SortOffers.propTypes = {
  activeSortType: PropTypes.string,
  isOpen: PropTypes.bool,
  toggleSort: PropTypes.func,
  setSortType: PropTypes.func,
  sortFilteredOffers: PropTypes.func
};

export default connect(
  (state) => ({
    activeSortType: getActiveSortType(state),
    isOpen: getIsSortOpen(state),
  }),
  (dispatch) => ({
    toggleSort: () => dispatch(ActionCreator.toggleSort()),
    setSortType: (sortType) => dispatch(ActionCreator.setSortType(sortType)),
    sortFilteredOffers: () => dispatch(ActionCreator.sortFilteredOffers()),
  })
)(SortOffers);
