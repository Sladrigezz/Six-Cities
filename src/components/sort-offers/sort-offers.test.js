import React from 'react';
import renderer from 'react-test-renderer';
import {SortOffers} from './sort-offers.jsx';

it(`should match snapshot`, () => {
  const toggleSort = jest.fn();
  const setSortType = jest.fn();
  const sortFilteredOffers = jest.fn();

  const props = {
    activeSortType: `POPULAR`,
    isOpen: true,
    toggleSort,
    setSortType,
    sortFilteredOffers
  };

  const wrapper = renderer.create(<SortOffers {...props} />).toJSON();

  expect(wrapper).toMatchSnapshot();
});
