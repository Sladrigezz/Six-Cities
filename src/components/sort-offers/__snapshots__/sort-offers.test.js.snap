// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<form
  action="#"
  className="places__sorting"
  method="get"
>
  <span
    className="places__sorting-caption"
  >
    Sort by
  </span>
  <span
    className="places__sorting-type"
    onClick={[Function]}
    tabIndex="0"
  >
    Popular
    <svg
      className="places__sorting-arrow"
      height="4"
      width="7"
    >
      <use
        xlinkHref="#icon-arrow-select"
      />
    </svg>
  </span>
  <ul
    className="places__options places__options--custom places__options--opened"
  >
    <li
      className="places__option places__option--active"
      id="POPULAR"
      onClick={[Function]}
      tabIndex={0}
    >
      Popular
    </li>
    <li
      className="places__option "
      id="PRICEASC"
      onClick={[Function]}
      tabIndex={1}
    >
      Price: low to high
    </li>
    <li
      className="places__option "
      id="PRICEDESC"
      onClick={[Function]}
      tabIndex={2}
    >
      Price: high to low
    </li>
    <li
      className="places__option "
      id="RATEDESC"
      onClick={[Function]}
      tabIndex={3}
    >
      Top rated first
    </li>
  </ul>
</form>
`;
